doc/Witch/intro-to-witch.md

```diff
@@ -0,0 +1,127 @@
+    automate the process (e.g, proof search, SMT solver) may be
+    non-deterministic or never find a solution. Therefore, the system
+    should be able to handle "impure" computations and errors. Algebraic
+    effects and handlers have well-defined semantics and provide a
+    simple interface for performing effects. With them, we avoid
+    indiscriminate effects that are often error-prone and intricate
+    effects machinery such as monad transformers;
+
+2.  the semantics of effects and handlers accommodate composition of
+    arbitrary effects and the composition of multiple handlers, which
+    means users have the ability to weaken more general strategies into
+    specific ones while maintaining top-level handlers unmodified;
+
+3.  effects and handlers have well-defined semantics, and it is the main
+    feature of many new languages, such as
+    [Koka](https://koka-lang.github.io/koka/doc/index.html),
+    [Eff](https://www.eff-lang.org), [Frank](https://github.com/frank-lang) and 
+    [Effekt](https://effekt-lang.org), and have implementation in all major
+    functional programming languages, what guarantees that our approach is based
+    in research and real-world experiences, and not something we hacked together
+    over the weekend. 
+
+## The Essence of Witch
+
+As for the syntax, we use `operation { params } via handler`, which is
+a syntactic improvement over `handler(operation, params)`, since effect handling
+is similar to function application, but also carries effect information.
+The user defines data types and functions as usual, and then uses
+Witch to prove properties about said definitions. The Examples below
+shows it in action.
+
+![](paper/image/witch.svg)
+
+The proof of commutativity of addition under
+natural numbers and of associativity of list concatenation are shown,
+and use the three main effects: Solver, Proof Search and Tactics. In the
+proof assistant literature, there exists no precise definition of
+commonly used terms "solver", "proof search" and "tactics". All these
+terms are used in different communities, and mean some variation of
+"automatic strategy to construct a term under certain constraints".
+
+
+In Witch, however, we use the following definitions of previously known terms:
+
+-   The Solver effect is used for integration with external solvers via
+    IO; we believe should suffice for the user to write, e.g. `SMT`, and
+    handlers should implement internal strategies to choose between the
+    different theories supported by solvers. If the black-box approach
+    to solvers presents itself a challenge, a specialisation of the
+    handler is possible, e.g. `operation { params } via Z3.QF-UFLIA`
+    .
+
+-   The Proof Search effect is used for library-level algorithms; users
+    may choose to implement their own algorithms using a limited set of
+    meta-programming[^1] constructs that are handled at top-level[^2].
```

Comment:
```suggestion
    meta-programming<sup>1</sup> constructs that are handled at top-level<sup>2</sup>.
```
Maybe making these superscripts will look better?

Reply:
That's how GitHub-flavoured markdown does footnotes, we can change what's within brackets, but there's nothing else we can do.